### PR TITLE
hack for proto in case of many-repo with relative package

### PR DESCRIPTION
**Pull request by natansil**
This hack solves an issue with proto_library target depending on proto_library from external repo.
`scalapb_proto_library` failed to find the proto from the external repo.

This hack can be reverted when https://github.com/bazelbuild/bazel/issues/4665 is resolved.

### CODE REVIEW

File: scala_proto/scala_proto.bzl [hunk 1]
```diff
@@ -331,14 +331,28 @@ def scala_proto_repositories(
       name = 'io_bazel_rules_scala/dependency/proto/netty_handler_proxy',
       actual = '@scala_proto_rules_netty_handler_proxy//jar')
 
-def _root_path(f):
+def _root_path(f, transitive_proto_paths):
+#   print("f.is_source is {}".format(f.is_source))
```
Comment (ittaiz) [inline scala_proto/scala_proto.bzl]: Please remove all prints

File: src/scala/scripts/PBGenerateRequest.scala [hunk 2]
```diff
@@ -10,13 +10,23 @@ object PBGenerateRequest {
...
+        println("parsed(0) is " + parsed(0))
+        println("parsed(1) is " + parsed(1))
+        (parsed(0), parsed(1))
+      }
+      else {
+        val root = parsed(0)
+        val file = if (root.isEmpty) {
+          parsed(1)
+        } else {
+          parsed(1).substring(root.length + 1)
```
[the remark this answers is not in the document]
Comment (natansil) [reply src/scala/scripts/PBGenerateRequest.scala]: Note this line is under condition where there is no external in root...

File: scala_proto/scala_proto.bzl [hunk 3]
```diff
@@ -331,14 +331,28 @@ def scala_proto_repositories(
...
+        return relative_path
+    else:
+        return f.owner.workspace_root
   return '/'.join([f.root.path, f.owner.workspace_root])
 
-def _colon_paths(data):
+def _colon_paths(data, transitive_proto_paths):
   return ':'.join([
-      "{root},{path}".format(root = _root_path(f), path = f.path)
+      "{root},{path}".format(root = _root_path(f, transitive_proto_paths), path = f.path)
```
[the remark this answers is not in the document]
Comment (natansil) [reply scala_proto/scala_proto.bzl]: In rules_scala this pattern of mixed scala and Starlark is quite common...

File: scala_proto/scala_proto.bzl [hunk 4]
```diff
@@ -331,14 +331,28 @@ def scala_proto_repositories(
...
   if f.is_source:
-    return f.owner.workspace_root
+    print("f.path: " + f.path + ".  f.owner.workspace_root: "+ f.owner.workspace_root + ". f.root.path: " + f.root.path)
+    if (f.owner.workspace_root):
+        start = len(f.owner.workspace_root)+1
+        # print("start index: {}".format(start))
+        package_len = 0
+        for package_set in transitive_proto_paths:
+           for package in package_set.to_list():
+               if str(package) in f.path:
```
[the remark this answers is not in the document]
Comment (natansil) [reply scala_proto/scala_proto.bzl]: ok, I'll do something like:
```suggestion
               if f.path.startswith(f.owner.workspace_root + "/" + str(package)):
```

File: src/scala/scripts/PBGenerateRequest.scala [hunk 5]
```diff
@@ -10,13 +10,23 @@ object PBGenerateRequest {
     val jarOutput = args.get(0)
     val parsedProtoFiles = args.get(1).split(':').toList.map { rootAndFile =>
       val parsed = rootAndFile.split(',')
-      val root = parsed(0)
-      val file = if (root.isEmpty) {
-        parsed(1)
-      } else {
-        parsed(1).substring(root.length + 1)
+      if (parsed(1).contains("external")) {
```
Comment (natansil) [inline src/scala/scripts/PBGenerateRequest.scala]: @ittaiz, I've found the bug manifested in OriP's branch. This line had the error.
changing it to `startsWith`:
```suggestion
      if (parsed(1).startsWith("external")) {
```

fixed it.

next step is to add E2E test.
should we do this in Wix's fork or on bazelbuild?

**Comment by ittaiz** [comment]
@natansil can we close this?

**Pull request closed without merging**